### PR TITLE
Migrate to Hardhat + Waffle

### PR DESCRIPTION
**Pull request by krebernisak**
- Update OZ lib to v3.4.0, polish contracts
- Update to Ethers v5 & Typechain v4
- Test refactored to share code by reusing the same behavior implementation, assertion simplified using Waffle
- To support Go code generation from build artifacts we remove the metadata from the compiled bytecode.
- Update to version 1.2.0-alpha
- Add contract compiled artifacts to published package
- Expose methods like `getContractFactory` to find compiled artifacts that consumers might find useful.

### CODE REVIEW

File: hardhat.config.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  optimizer,
+  metadata: {
+    // To support Go code generation from build artifacts
+    // we need to remove the metadata from the compiled bytecode.
+    bytecodeHash: 'none',
+  },
+}
+
+const versions: Record<string, SolcConfig> = {
+  'v0.4': { version: '0.4.16', settings },
```
Comment (alexroan) [inline hardhat.config.ts:25]: This version doesn't work for me locally.
```
Downloading compiler 0.4.16
Error HH503: Couldn't download compiler version 0.4.16. Checksum verification failed. Please check your connection.
```
0.4.24 works fine though.

File: hardhat.config.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import { HardhatUserConfig, SolcConfig } from 'hardhat/types'
+import { hardhat } from './src'
+
+import '@nomiclabs/hardhat-waffle'
+import '@typechain/hardhat'
+
+const DEFAULT_VERSION = 'v0.6'
+
+const optimizer = {
+  runs: 200,
```
Comment (alexroan) [inline hardhat.config.ts:11]: Do we not optimize a crazy high amount like `10000000` or something?

File: package.json [hunk 3]
```diff
@@ -2,41 +2,51 @@
...
-    "test:v0.4": "env NODE_ENV=test find ./test/v0.4 -type f | xargs truffle test",
-    "test:v0.6": "env NODE_ENV=test find ./test/v0.6 -type f | xargs truffle test",
-    "migrate:v0.4": "env VERSION=v0.4 truffle migrate",
-    "migrate:v0.6": "env VERSION=v0.6 truffle migrate"
+    "clean": "tsc -b --clean tsconfig.json && rm -rf build/",
+    "setup": "yarn build:contracts",
+    "build:contracts": "yarn build:contracts:v0.4 && yarn build:contracts:v0.6",
+    "build:contracts:v0.4": "VERSION=v0.4 hardhat compile --show-stack-traces",
+    "build:contracts:v0.6": "VERSION=v0.6 hardhat compile --show-stack-traces",
+    "test": "hardhat test --show-stack-traces"
```
Comment (alexroan) [inline package.json:23]: We can add a simple Github CI workflow that runs yarn install, setup and test now :)

**Review by alexroan — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (krebernisak) [reply package.json:23]: Yes, but in another PR :) The work on this started [here](https://github.com/smartcontractkit/LinkToken/pull/29/commits/346e1598a4f5337b56fc71b81b8c0be0c2c54fcd).

File: hardhat.config.ts (hunk 2 above)
Comment (krebernisak) [reply hardhat.config.ts:11]: The LinkToken v4 was at 200, so I continued with 200 to get the same compiler output as the contract is already deployed. v6 is in a similar situation but going forward we could raise the number for v7 for example.

File: hardhat.config.ts (hunk 1 above)
Comment (krebernisak) [reply hardhat.config.ts:25]: Hmm, works on my machine...

`0.4.16` was the version the original LinkToken v4 was compiled with.

File: hardhat.config.ts (hunk 2 above)
Comment (krebernisak) [reply hardhat.config.ts:11]: I've added this at ["Use 1_000_000 optimizer runs for v0.7"](https://github.com/smartcontractkit/LinkToken/pull/29/commits/8c90ff9dab891e4cd79937790f795eefb0f58b84). Left 200 for older versions.

File: hardhat.config.ts (hunk 1 above)
Comment (se3000) [reply hardhat.config.ts:25]: Worked for me.

File: test/behavior/ERC677Token.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
+import { ethers } from 'hardhat'
+import { expect } from 'chai'
+import { ContractFactory, Contract, Signer } from 'ethers'
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
+
+import * as h from '../helpers'
+
+export const shouldBehaveLikeERC677Token = (
```
Comment (se3000) [inline test/behavior/ERC677Token.ts:8]: Very excited for shared behavior specs.

File: tsconfig.base.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
+{
+  "compilerOptions": {
+    "outDir": "dist",
+    /* Basic Options */
```
Comment (se3000) [inline tsconfig.base.json:4]: What is going on with this syntax highlighting? I feel like I'm being punished.

**Review by se3000 — APPROVED** [review]

**Pull request merged**